sway-core/src/language/ty/declaration/constant.rs

```diff
@@ -29,7 +31,15 @@ impl PartialEqWithEngines for TyConstantDeclaration {
             && type_engine
                 .get(self.return_type)
                 .eq(&type_engine.get(other.return_type), engines)
-            && self.attributes == other.attributes
-            && self.span == other.span
+    }
+}
+
+impl HashWithEngines for TyConstantDeclaration {
+    fn hash<H: Hasher>(&self, state: &mut H, engines: Engines<'_>) {
+        let type_engine = engines.te();
+        self.name.hash(state);
+        self.value.hash(state, engines);
+        self.visibility.hash(state);
+        type_engine.get(self.return_type).hash(state, engines);
```

Comment:
Should we also hash `is_configurable` here?

Reply:
In fact, with all of these custom implementations of `hash`, how do we decide what fields should be hashed? And how do we make sure that this is future proof when new fields are added? (hopefully this is rare but `is_configurable` is an example of a new field that may have been missed, even in `eq` above).

Reply:
> how do we decide what fields should be hashed?

There is not any perfect science to it... although there should be. In general there is poor separation between the information needed for static analysis + type checking (only some stuff should be hashed) and the information needed for the consumers of the compiler, LSP, etc, (other stuff should to be hashed). The implementations of `Hash` and such have historically accommodated the former case, leading to workarounds like [`BaseIdent`](https://github.com/FuelLabs/sway/commit/80a8d1b7382fd70a7e417c30a94d30778d56e762) being introduced.

Regarding this larger problem, I'm not yet sure what the best course of action should be. I'll need to do some research to learn about different approaches and the current landscape.

> And how do we make sure that this is future proof when new fields are added?

So for what to do right now, I propose that I change this PR slightly so that adding new fields will give a Rust error unless you decide if the new field should be hashed or not. And I'll add little explanation blurbs for the fields left out. What are your thoughts?

Reply:
Your proposal seems reasonable to me. And since we using a custom hash trait instead of the standard `Hash`, maybe we need two separate traits with two separate implementations to be used in two different scenarios. One would call the other.